wp/lib/bap_wp/src/precondition.ml

```diff
@@ -1289,16 +1289,39 @@ let user_func_spec
   debug "Making user-defined subroutine spec with subroutine-name: %s, pre:
 %s, post: %s \n%!" sub_name sub_pre sub_post;
   if String.equal sub_name (Sub.name sub) then
-    let spec env post tid = (
+    let spec env post _ = (
       (* turn strings into proper smtlib2 statements; incr stack_ptr *)
-      let sub_pre : Constr.t = Z3_utils.mk_smtlib2_single env sub_pre in
-      let sub_post : Constr.t = Z3_utils.mk_smtlib2_single env sub_post in
+      let (sub_pre, pre_init_vars, pre_vars)
+          : Constr.t  * string list * string list =
+        Z3_utils.mk_smtlib2_single_with_vars env sub_pre in
+      let (sub_post, post_init_vars, post_vars)
+          : Constr.t * string list * string list =
+        Z3_utils.mk_smtlib2_single_with_vars env sub_post in
       let sub_post, env = increment_stack_ptr sub_post env in
-      (* collect (physical) inputs/outputs of sub *)
-      let sub_inputs : Var.t list = vars_from_sub env sub |> Var.Set.to_list in
-      let sub_inputs =
-        List.filter sub_inputs ~f:(fun v -> Var.is_physical v) in
-      let sub_outputs : Var.t list = sub_inputs in
+      (* TODO: Some sanity checking on sub_pre and sub_post would be nice
+         here.  If you just forget the word "assert" you silently get some
+         kind of empty constr. *)
+
+      (* Next, collect inputs/outputs of sub.  We do this by constructing a set
+         of all potential inputs/outputs (all the registers and mem), then
+         filtering out everything not mentioned by the user-provided specs.
+
+         Question: is SP + GPRs really all the registers?  *)
+      let potential_ios : Bap.Std.Var.Set.t =
```

Comment:
This was the reason why we looked at the subroutine, and found the union of SP, the GPRs, and the vars found in the sub just in case we were missing registers. It seems we can't do that here since we won't always have the body of the subroutine. I wonder if http://binaryanalysisplatform.github.io/bap/api/master/bap-core-theory/Bap_core_theory/Theory/Target/index.html#val-vars gives us something different.

Reply:
Good idea.  I tested this and it returns everything we were already getting, plus some flag registers.  So I added a function to `environment.ml` like `get_gprs` and used it here.